src/servers/src/http/authorize.rs

```diff
@@ -70,45 +71,86 @@ where
             };
 
-            let (scheme, credential) = match auth_header(&request) {
-                Ok(auth_header) => auth_header,
+            // do authenticate
+            match authenticate(&user_provider, &request).await {
+                Ok(user_info) => {
+                    request.extensions_mut().insert(user_info);
+                }
                 Err(e) => {
-                    error!("failed to get http authorize header, err: {:?}", e);
+                    error!("authenticate failed: {}", e);
                     return Err(unauthorized_resp());
                 }
-            };
+            }
 
-            match scheme {
-                AuthScheme::Basic => {
-                    let (username, password) = match decode_basic(credential) {
-                        Ok(basic_auth) => basic_auth,
-                        Err(e) => {
-                            error!("failed to decode basic authorize, err: {:?}", e);
-                            return Err(unauthorized_resp());
-                        }
-                    };
-                    match user_provider
-                        .auth(
-                            Identity::UserId(&username, None),
-                            crate::auth::Password::PlainText(&password),
-                        )
-                        .await
-                    {
-                        Ok(user_info) => {
-                            request.extensions_mut().insert(user_info);
-                            Ok(request)
-                        }
-                        Err(e) => {
-                            error!("failed to auth, err: {:?}", e);
-                            Err(unauthorized_resp())
-                        }
-                    }
+            match authorize(&user_provider, &request).await {
+                Ok(_) => Ok(request),
+                Err(e) => {
+                    error!("authorize failed: {}", e);
+                    Err(unauthorized_resp())
                 }
             }
         })
     }
 }
 
+async fn authorize<B: Send + Sync + 'static>(
+    user_provider: &UserProviderRef,
+    request: &Request<B>,
+) -> crate::auth::Result<()> {
+    // try get database name
+    let query = request.uri().query().unwrap_or_default();
+    let input_database = match serde_urlencoded::from_str::<HashMap<String, String>>(query) {
+        Ok(query_map) => query_map
+            .get("db")
+            .context(IllegalParamSnafu {
+                msg: "fail to get valid database from http query",
+            })?
+            .to_owned(),
+        Err(e) => IllegalParamSnafu {
+            msg: format!("fail to parse http query: {e}"),
+        }
+        .fail()?,
+    };
+
+    let (catalog, database) =
+        crate::parse_catalog_and_schema_from_client_database_name(&input_database);
+
+    let user_info = request
+        .extensions()
+        .get::<UserInfo>()
+        .context(IllegalParamSnafu {
```

Comment:
You can use `InternalSnafu`. Or better, create an `InvalidStateSnafu` with status code "unexpected".